### ISSUE TITLE
Multiple calls of the Read Field Type Policy which process the previously processed value

### ISSUE THREAD

**Issue by FOODy** [apollographql/apollo-client#11227]
### Issue Description

Hello,

i have encountered a bug in the handling of the "read" field type policy.

The problem seems to be when:
 1. there is a read field type policy defined
 2. multiple (cache-wise matching) queries start fetching at the "same" time

They seem to corrupt their results by calling the read policy another time on a value that has already been processed by the policy.

### Link to Reproduction

https://github.com/FOODy/apollo-client-cache-typepolicy-read-multiple-call-bug

### Reproduction Steps

Run the example and take a look at the names in the person list.

After the person name should be only one "+", but because of the explained bug you will see "++".

Here are the relevant code snippets:

```tsx

const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      Person: {
        fields: {
          name: {
            read: (existing) => existing != null ? existing + '+' : existing
          }
        },
      }
    }
  }),
  link,
});

```

```tsx
function App() {
// ....
  const { loading, data } = useQuery(ALL_PEOPLE);
  const _another = useQuery(ALL_PEOPLE); // <- Triggers another call on read
// ....
```

Wrong Result:

![image](https://github.com/apollographql/apollo-client/assets/151868/d1bc6014-082d-4d5d-bf23-aa96c5733793)

As i understand, the read policy should *not* modify the cache itself.

**Comment by bignimbus** [comment]
Hi @FOODy 👋🏻 thanks for opening this issue!  I'm able to reproduce this on my end and will be investigating with the team hopefully next week.  I'll let you know what we find!

**Comment by kekami** [comment]
@bignimbus Has there been any progress on this?

**Comment by bignimbus** [comment]
Thanks for the nudge @kekami!  I do have some updates for you and @FOODy:

@FOODy's understanding of the expected behavior of `read` is correct.  This was indeed a bug and was resolved by https://github.com/apollographql/apollo-client/pull/11202, which is available as of `3.9.0-alpha.0` - you can try it out via `npm i @apollo/client@next`.  You'll notice, however, that the linked reproduction still has some bugs after upgrading to the latest 3.9 Alpha version.  This is due to the usage of `writeQuery` here:

https://github.com/FOODy/apollo-client-cache-typepolicy-read-multiple-call-bug/blob/067a1ac5094a5d97680b14009ce4f5b29213331d/src/index.jsx#L44-L53

Notice that we're accumulating `"+"` due to reading the field and then writing the `"+"` directly into the cache.

Does that make sense?  Try upgrading and let us know whether it resolves the issue in your app.  Let me know if you have any more questions!
